### PR TITLE
feat: awaitable connect

### PR DESCRIPTION
**Pull request by paulpach**
You can now await for the transport client to connect.  This makes it a lot easier to capture errors and know when the connection is established

BREAKING CHANGE: ConnectClient has been replaced by ConnectClientAsync

### CODE REVIEW

**Comment by paulpach** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 4
- Added 2
           

Complexity increasing per file
==============================
- Assets/Mirror/Runtime/Transport/Telepathy/Client.cs  1
         

Complexity decreasing per file
==============================
+ Assets/Mirror/Examples/ListServer/ListServer.cs  -1
         

Clones added
============
- Assets/Mirror/Tests/Editor/Telepathy/TransportTest.cs  2
         

Clones removed
==============
+ Assets/Mirror/Tests/Editor/FallbackTest.cs  -1
+ Assets/Mirror/Tests/Editor/FallbackTransportTest.cs  -1
+ Assets/Mirror/Tests/Editor/MultiplexTest.cs  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/paulpach/MirrorNG/pullRequest?prid=4917397&bid=16273489)

**Comment by Lymdun** [comment]
Nice PR, but why did you modified ServerDisconnect too?

**Comment by paulpach** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 6
- Added 26
           

Complexity increasing per file
==============================
- Assets/Mirror/Tests/Editor/Tcp/TcpTest.cs  3
- Assets/Mirror/Runtime/Transport/Telepathy/Client.cs  1
         

Complexity decreasing per file
==============================
+ Assets/Mirror/Examples/ListServer/ListServer.cs  -1
+ Assets/Mirror/Runtime/Transport/Tcp/Client.cs  -1
         

Clones added
============
- Assets/Mirror/Tests/Editor/Tcp/TcpTest.cs  8
- Assets/Mirror/Tests/Editor/Telepathy/TransportTest.cs  4
         

Clones removed
==============
+ Assets/Mirror/Tests/Editor/FallbackTest.cs  -1
+ Assets/Mirror/Tests/Editor/MultiplexTest.cs  -1
+ Assets/Mirror/Tests/Editor/FallbackTransportTest.cs  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/paulpach/MirrorNG/pullRequest?prid=4917397&bid=16273489)

File: Assets/Mirror/Runtime/Transport/Tcp/Server.cs [hunk 1]
```diff
@@ -32,7 +32,7 @@ public class Server : Common
         // (right now we only use it from one listener thread, but we might have
         //  multiple threads later in case of WebSockets etc.)
         //  HLAPI uses 0 for local connection,  so our ids start with 1
-        int counter = 1;
+        int counter = 0;
```
Comment (Lymdun) [inline Assets/Mirror/Runtime/Transport/Tcp/Server.cs]: Please explain/fix the above comment

**Comment by paulpach** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 6
- Added 26
           

Complexity increasing per file
==============================
- Assets/Mirror/Tests/Editor/Tcp/TcpTest.cs  3
- Assets/Mirror/Runtime/Transport/Telepathy/Client.cs  1
         

Complexity decreasing per file
==============================
+ Assets/Mirror/Runtime/Transport/Tcp/Client.cs  -1
+ Assets/Mirror/Examples/ListServer/ListServer.cs  -1
         

Clones added
============
- Assets/Mirror/Tests/Editor/Tcp/TcpTest.cs  8
- Assets/Mirror/Tests/Editor/Telepathy/TransportTest.cs  4
         

Clones removed
==============
+ Assets/Mirror/Tests/Editor/FallbackTest.cs  -1
+ Assets/Mirror/Tests/Editor/FallbackTransportTest.cs  -1
+ Assets/Mirror/Tests/Editor/MultiplexTest.cs  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/paulpach/MirrorNG/pullRequest?prid=4917397&bid=16273489)

**Comment by paulpach** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 6
- Added 26
           

Complexity increasing per file
==============================
- Assets/Mirror/Runtime/Transport/Telepathy/Client.cs  1
- Assets/Mirror/Tests/Editor/Tcp/TcpTest.cs  3
         

Complexity decreasing per file
==============================
+ Assets/Mirror/Examples/ListServer/ListServer.cs  -1
+ Assets/Mirror/Runtime/Transport/Tcp/Client.cs  -1
         

Clones added
============
- Assets/Mirror/Tests/Editor/Tcp/TcpTest.cs  8
- Assets/Mirror/Tests/Editor/Telepathy/TransportTest.cs  4
         

Clones removed
==============
+ Assets/Mirror/Tests/Editor/FallbackTransportTest.cs  -1
+ Assets/Mirror/Tests/Editor/FallbackTest.cs  -1
+ Assets/Mirror/Tests/Editor/MultiplexTest.cs  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/paulpach/MirrorNG/pullRequest?prid=4917397&bid=16273489)

File: Assets/Mirror/Runtime/NetworkServer.cs [hunk 2]
```diff
@@ -34,7 +34,7 @@ public class NetworkServer : MonoBehaviour
         /// <summary>
         /// True if there is a local client connected to this server (host mode)
         /// </summary>
-        public bool LocalClientActive => localClient != null & localClient.active;
+        public bool LocalClientActive => localClient != null && localClient.active;
```
Comment (Lymdun) [inline Assets/Mirror/Runtime/NetworkServer.cs]: Wow, should be fixed in master

**Comment by paulpach** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 6
- Added 26
           

Complexity increasing per file
==============================
- Assets/Mirror/Runtime/Transport/Telepathy/Client.cs  1
- Assets/Mirror/Tests/Editor/Tcp/TcpTest.cs  3
         

Complexity decreasing per file
==============================
+ Assets/Mirror/Runtime/Transport/Tcp/Client.cs  -1
+ Assets/Mirror/Examples/ListServer/ListServer.cs  -1
         

Clones added
============
- Assets/Mirror/Tests/Editor/Telepathy/TransportTest.cs  4
- Assets/Mirror/Tests/Editor/Tcp/TcpTest.cs  8
         

Clones removed
==============
+ Assets/Mirror/Tests/Editor/FallbackTransportTest.cs  -1
+ Assets/Mirror/Tests/Editor/FallbackTest.cs  -1
+ Assets/Mirror/Tests/Editor/MultiplexTest.cs  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/paulpach/MirrorNG/pullRequest?prid=4917397&bid=16273489)

File: Assets/Mirror/Runtime/NetworkServer.cs (hunk 2 above)
Comment (paulpach) [reply Assets/Mirror/Runtime/NetworkServer.cs]: yes.  I fixed it in master now

**Comment by paulpach** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 6
- Added 26
           

Complexity increasing per file
==============================
- Assets/Mirror/Runtime/Transport/Telepathy/Client.cs  1
- Assets/Mirror/Tests/Editor/Tcp/TcpTest.cs  3
         

Complexity decreasing per file
==============================
+ Assets/Mirror/Runtime/Transport/Tcp/Client.cs  -1
+ Assets/Mirror/Examples/ListServer/ListServer.cs  -1
         

Clones added
============
- Assets/Mirror/Tests/Editor/Tcp/TcpTest.cs  8
- Assets/Mirror/Tests/Editor/Telepathy/TransportTest.cs  4
         

Clones removed
==============
+ Assets/Mirror/Tests/Editor/FallbackTransportTest.cs  -1
+ Assets/Mirror/Tests/Editor/FallbackTest.cs  -1
+ Assets/Mirror/Tests/Editor/MultiplexTest.cs  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/paulpach/MirrorNG/pullRequest?prid=4917397&bid=16273489)

**Comment by paulpach** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 6
- Added 27
           

Complexity increasing per file
==============================
- Assets/Mirror/Tests/Editor/Tcp/TcpTest.cs  3
- Assets/Mirror/Runtime/Transport/Tcp/Server.cs  1
- Assets/Mirror/Runtime/Transport/Telepathy/Client.cs  1
         

Complexity decreasing per file
==============================
+ Assets/Mirror/Examples/ListServer/ListServer.cs  -1
+ Assets/Mirror/Runtime/Transport/Tcp/Client.cs  -1
         

Clones added
============
- Assets/Mirror/Tests/Editor/Tcp/TcpTest.cs  8
- Assets/Mirror/Tests/Editor/Telepathy/TransportTest.cs  4
         

Clones removed
==============
+ Assets/Mirror/Tests/Editor/MultiplexTest.cs  -1
+ Assets/Mirror/Tests/Editor/FallbackTest.cs  -1
+ Assets/Mirror/Tests/Editor/FallbackTransportTest.cs  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/paulpach/MirrorNG/pullRequest?prid=4917397&bid=16273489)

**Comment by paulpach** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 6
- Added 26
           

Complexity increasing per file
==============================
- Assets/Mirror/Tests/Editor/Tcp/TcpTest.cs  3
- Assets/Mirror/Runtime/Transport/Tcp/Server.cs  1
- Assets/Mirror/Runtime/Transport/Telepathy/Client.cs  1
         

Complexity decreasing per file
==============================
+ Assets/Mirror/Examples/ListServer/ListServer.cs  -1
+ Assets/Mirror/Runtime/Transport/Tcp/Client.cs  -1
         

Clones added
============
- Assets/Mirror/Tests/Editor/Tcp/TcpTest.cs  8
- Assets/Mirror/Tests/Editor/Telepathy/TransportTest.cs  4
         

Clones removed
==============
+ Assets/Mirror/Tests/Editor/MultiplexTest.cs  -1
+ Assets/Mirror/Tests/Editor/FallbackTransportTest.cs  -1
+ Assets/Mirror/Tests/Editor/FallbackTest.cs  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/paulpach/MirrorNG/pullRequest?prid=4917397&bid=16273489)

**Comment by paulpach** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 6
- Added 25
           

Complexity increasing per file
==============================
- Assets/Mirror/Tests/Editor/Tcp/TcpTest.cs  3
- Assets/Mirror/Runtime/Transport/Tcp/Server.cs  1
- Assets/Mirror/Runtime/Transport/Telepathy/Client.cs  1
         

Complexity decreasing per file
==============================
+ Assets/Mirror/Runtime/Transport/Tcp/Client.cs  -1
+ Assets/Mirror/Examples/ListServer/ListServer.cs  -1
         

Clones added
============
- Assets/Mirror/Runtime/Transport/Tcp/TcpTransport.cs  1
- Assets/Mirror/Tests/Editor/Tcp/TcpTest.cs  8
- Assets/Mirror/Tests/Editor/Telepathy/TransportTest.cs  4
- Assets/Mirror/Runtime/Transport/TelepathyTransport.cs  1
         

Clones removed
==============
+ Assets/Mirror/Tests/Editor/FallbackTransportTest.cs  -1
+ Assets/Mirror/Tests/Editor/MultiplexTest.cs  -1
+ Assets/Mirror/Tests/Editor/FallbackTest.cs  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/paulpach/MirrorNG/pullRequest?prid=4917397&bid=16273489)

**Pull request merged**